JAVA-2255: Add annotation to indicate a property is transient

tolbertam opened the pull request:
For [JAVA-2255](https://datastax-oss.atlassian.net/browse/JAVA-2255)

---

Motivation:

It is sometimes desirable for properties of an entity to not be
considered by an object mapper.

Modifications:

Introduced a `@Transient` annotation that when applied to a field or
getter, causes the mapper to not include this field for generated
queries.

In addition, if a field includes the 'transient' keyword, it is also not
considered.

Result:

properties including a `@Transient` annotation or having a field with
transient keyword will not be considered by the mapper.

tolbertam committed:
JAVA-2255: Add annotation to indicate a property is transient

Motivation:

It is sometimes desirable for properties of an entity to not be
considered by an object mapper.

Modifications:

Introduced a @Transient annotation that when applied to a field or
getter, causes the mapper to not include this field for generated
queries.

In addition, if a field includes the 'transient' keyword, it is also not
considered.

Result:

properties including a @Transient annotation or having a field with
transient keyword will not be considered by the mapper.

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -347,4 +352,17 @@ private CqlNameGenerator buildCqlNameGenerator(TypeElement classElement) {
   }
+
+  private boolean isTransient(ExecutableElement getMethod, VariableElement field) {
+    Transient getterAnnotation = getMethod.getAnnotation(Transient.class);
```
Of note: we don't check that other annotations, such as `PartitionKey` or `ClusteringKey` are present and emit a warning, is that worth doing?

olim7t replied:
Yes, if it's not too much work I think it would be worth it.

tolbertam replied:
:+1: will add that

tolbertam commented on mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/Transient.java:
```diff
@@ -0,0 +1,40 @@
+ */
+@Target({ElementType.FIELD, ElementType.METHOD})
+@Retention(RetentionPolicy.CLASS)
+public @interface Transient {}
```
Some things that the 3.x mapper has that this (currently) doesn't:

1. There was a way of specifying a [`PropertyTransienceStrategy`](https://github.com/datastax/java-driver/blob/3.x/driver-mapping/src/main/java/com/datastax/driver/mapping/PropertyTransienceStrategy.java), which allowed one to choose whether properties need to be explicitly annotated or not to be included.  Since we don't have a `@Column` annotation, this didn't seem appropriate here.

2. You could provide a global configuration for specifying transient property names, i.e.: [`addTransientPropertyNames`](https://github.com/datastax/java-driver/blob/3.x/driver-mapping/src/main/java/com/datastax/driver/mapping/DefaultPropertyMapper.java#L179-L242).  I think a good solution for this might be to add an entity-scoped annotation named `@TransientProperties` that is akin to jackson-annotations' [`JsonIgnoreProperties`](https://fasterxml.github.io/jackson-annotations/javadoc/2.6/com/fasterxml/jackson/annotation/JsonIgnoreProperties.html), i.e.:

```java
@TransientProperties({"a", "b", "c", "d"})
@Entity
class MyEntity {}
```

what do you think?

olim7t replied:
1. I guess the question is: do we need an `@Column` annotation? I'm not really convinced, mapping all by default sounds pretty good as long as there's a proper exclusion mechanism.
Also this can always be added later as long as it defaults to "map all", so no need to rush it.

2. That sound like a good idea, for those cases where you inherit from a class that you don't own, and which has properties that you don't want to map.

tolbertam replied:
1. Agreed that this isn't compelling enough to add a `@Column`.
2. :+1:, I'll add that.

tolbertam commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/TransientIT.java:
```diff
@@ -0,0 +1,288 @@
+import org.junit.rules.TestRule;
+
+@Category(ParallelizableTests.class)
+public class TransientIT {
```
I'm only using `Transient` here because that was what we used in 3.x.

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/TransientIT.java:
```diff
@@ -0,0 +1,288 @@
+  public void should_ignore_field_with_transient_annotated_field() {
+    EntityWithTransientAnnotatedFieldDao dao =
+        mapper.entityWithTransientAnnotatedFieldDao(
+            sessionRule.keyspace(), CqlIdentifier.fromCql("entity"));
```
Nice use of table parameterization 👍

olim7t reviewed (approved):
LGTM so far, will review again after you do the changes we discussed (improved warnings and entity-level exclusions).

tolbertam committed:
Add TransientProperties annotation, additional validation

* Adds a `@TransientProperties` annotation for specifying what fields to
  consider transient at an Entity-level, which will be more useful once
  inheritance is supported.

* Adds additional validation to ensure `@PartitionKey` & `@ClusteringKey`
  annotations aren't used on transient properties.

tolbertam commented:
Pushed c580333 which adds an entity-level `@TransientProperties` annotation and additional validation around using other annotations on properties that are considered transient.

tolbertam commented on manual/mapper/entities/README.md:
```diff
@@ -159,10 +159,21 @@ private int day;
 This information is used by some of the DAO method annotations; for example,
 [@Select](../daos/select/)'s default behavior is to generate a selection by primary key.
 
+#### Transient properties
```
Just realized I need to add doc for `@TransientProperties`, will do

tolbertam committed:
Add documentation for `@TransientProperties`

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -347,4 +352,80 @@ private CqlNameGenerator buildCqlNameGenerator(TypeElement classElement) {
+    Class<?> otherAnnotationClass =
+        partitionKeyIndex >= 0
+            ? PartitionKey.class
+            : clusteringColumnIndex >= 0 ? ClusteringColumn.class : null;
```
We should also check other mapper annotations: `CqlName`, `Computed` and maybe others in the future. I'm not sure if passing a value for each annotation will scale well, so maybe we could just re-read them on the field and getter.
```java
for (Class<?> annotationClass : ImmutableList.of(PartitionKey.class, ClusteringColumn.class, CqlName.class) {
  if (getMethod.getAnnotation(annotationClass) != null || field != null && field.getAnnotation(annotationClass) != null) {
    ...
  }
}
```
We could extract that into another method so that it's easier to call multiple times in this method (see my other comment).
```java
void checkTransientNotAnnotated(
      ExecutableElement getMethod,
      VariableElement field,
      Element elementToWarnOn)
```
Also possibly report all offending annotations if there is more than one.

tolbertam replied:
I like this idea, I think we could take it one step further and use it for more than just transient as I will also face the same issue with `@Computed`.  I realized that the 3.x mapper must have similar logic to prevent multiple conflicting annotations from being used on a property and that is the case: 

https://github.com/datastax/java-driver/blob/3.x/driver-mapping/src/main/java/com/datastax/driver/mapping/AnnotationChecks.java#L53-L93

I think the solution there works but is a bit more complicated than it needs to be.  I"ll try to come up with something.

tolbertam replied:
Actually, there are parts of it that I like a lot, such as resolving an annotation in a general way (i.e. pass both the field and the method to resolve the annotation), which will be nice for custom inheritence strategies when we add it.   I'll take a stab at refactoring things in a way to meet my needs, without trying to complicate things too much.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -347,4 +352,80 @@ private CqlNameGenerator buildCqlNameGenerator(TypeElement classElement) {
+    Transient getterAnnotation = getMethod.getAnnotation(Transient.class);
+    boolean isTransient;
+    Element transientElement = getMethod;
+    isTransient = getterAnnotation != null;
```
This is really hard to follow. How about treating the cases separately:
```java
if (getMethod.getAnnotation(Transient.class) != null) {
  checkTransientNotAnnotated(field, getMethod, getMethod);
  return true;
} else if (field != null && (field.getAnnotation(Transient.class) != null || field.getModifiers().contains(Modifier.TRANSIENT))) {
  checkTransientNotAnnotated(field, getMethod, field);
  return true;
}
return false;
```

And `checkTransientNotAnnotated` would have a more generic message "Transient properties can't be annotated with %s" to handle both field annotation and modifier.

tomekl007 replied:
Maybe we can go one step further and extract `field != null && (field.getAnnotation(Transient.class) != null || field.getModifiers().contains(Modifier.TRANSIENT)` condition to separete method to name that check (to make it more readable)

tolbertam committed:
Attempt to handle conflicting annotations in a consistent way

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -347,4 +304,126 @@ private CqlNameGenerator buildCqlNameGenerator(TypeElement classElement) {
+    return null;
+  }
+
+  private void scanFieldAnnotations(
```
code for `scanFieldAnnotations` and `scanMethodAnnotations` is largely the same, but anticipate that changing with [JAVA-2111](https://datastax-oss.atlassian.net/browse/JAVA-2111)

tolbertam commented:
looks like i have an errorprone issue to fix, will do

tolbertam committed:
Fix errorprone error, update test

tolbertam commented on mapper-processor/src/test/java/com/datastax/oss/driver/internal/mapper/processor/entity/EntityPropertyAnnotationsTest.java:
```diff
@@ -105,7 +105,7 @@ public void should_fail_with_expected_error(String expectedError, TypeSpec entit
     return new Object[][] {
       {
-        "Properties can't be annotated with both @PartitionKey and @ClusteringColumn.",
+        "Properties can't be annotated with both @ClusteringColumn and @PartitionKey.",
```
Was wondering if we had a way to test for compiler errors being surfaced, I should add some more tests as I had been testing manually up to this point, but if it expedites things more I can add tests when i work on [JAVA-2111](https://datastax-oss.atlassian.net/browse/JAVA-2111)

olim7t replied:
See `EntityPropertyAnnotationsTest`.

Those tests are pretty slow, so try not to add too many (I'm thinking of adding something in the POM to skip them).

olim7t replied:
Mmm forget that, looks like I misunderstood your question.

olim7t replied:
No need to expedite this PR, we might as well add all necessary tests now.

olim7t reviewed (approved):
OK, only nit is that it allows `Transient` + `CqlName`, which strictly speaking is not a valid combination, but we can live with that.

tolbertam commented:
> only nit is that it allows Transient + CqlName

I considered that case, but didn't seem too important to validate :+1:

tolbertam committed:
Add tests for Transient to EntityPropertyAnnotationsTest
